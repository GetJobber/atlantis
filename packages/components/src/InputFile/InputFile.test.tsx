import React from "react";
import { cleanup, fireEvent, render, waitFor } from "@testing-library/react";
import { InputFile } from ".";

afterEach(cleanup);

const testFile = new File(["ðŸ”± Atlantis"], "atlantis.png", {
  type: "image/png",
});

it("renders a InputFile", () => {
  const { container } = render(
    <InputFile getUploadParams={fetchUploadParams} />,
  );

  expect(container).toMatchSnapshot();
});

it("renders an InputFile with multiple uploads", () => {
  const { container } = render(
    <InputFile allowMultiple={true} getUploadParams={fetchUploadParams} />,
  );
  expect(container).toMatchSnapshot();
});

it("renders an InputFile with only images allowed", () => {
  const { container } = render(
    <InputFile allowedTypes="images" getUploadParams={fetchUploadParams} />,
  );
  expect(container).toMatchSnapshot();
});

it("renders an InputFile with multiple images allowed", () => {
  const { container } = render(
    <InputFile
      allowedTypes="images"
      allowMultiple={true}
      getUploadParams={fetchUploadParams}
    />,
  );
  expect(container).toMatchSnapshot();
});

it("renders an InputFile with proper variations", () => {
  const { container } = render(
    <>
      <InputFile getUploadParams={fetchUploadParams} />
      <InputFile size="small" getUploadParams={fetchUploadParams} />
      <InputFile variation="button" getUploadParams={fetchUploadParams} />
      <InputFile
        variation="button"
        size="small"
        getUploadParams={fetchUploadParams}
      />
    </>,
  );
  expect(container).toMatchSnapshot();
});

it("supports custom button label", () => {
  const buttonLabel = "Custom Label";
  const { getByText } = render(
    <InputFile buttonLabel={buttonLabel} getUploadParams={fetchUploadParams} />,
  );
  expect(getByText(buttonLabel)).toBeInstanceOf(HTMLElement);
});

it("properly notifies upload callbacks", async () => {
  const fetchParams = jest.fn(fetchUploadParams);
  const handleStart = jest.fn();
  const handleProgress = jest.fn();
  const handleComplete = jest.fn();

  const { container } = render(
    <InputFile
      getUploadParams={fetchParams}
      onUploadStart={handleStart}
      onUploadProgress={handleProgress}
      onUploadComplete={handleComplete}
    />,
  );
  const input = container.querySelector("input[type=file]");

  fireEvent.change(input, { target: { files: [testFile] } });

  const baseMatch = {
    key: "atlantis.png",
    name: "atlantis.png",
    size: expect.any(Number),
    progress: expect.any(Number),
    src: expect.any(Function),
    type: "image/png",
  };

  await waitFor(() => {
    expect(fetchParams).toHaveBeenCalledTimes(1);

    expect(handleStart).toHaveBeenCalledWith({
      ...baseMatch,
      progress: 0,
    });
    expect(handleProgress).toHaveBeenCalledWith(baseMatch);
    expect(handleComplete).toHaveBeenCalledWith({
      ...baseMatch,
      progress: 1,
    });

    expect(handleStart).toHaveBeenCalledTimes(1);
    expect(handleComplete).toHaveBeenCalledTimes(1);
  });
});

function fetchUploadParams(file: File) {
  return Promise.resolve({
    key: file.name,
    url: "https://httpbin.org/post",
    fields: { secret: "ðŸ¤«" },
  });
}
