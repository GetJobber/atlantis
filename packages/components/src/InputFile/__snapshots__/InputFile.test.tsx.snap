// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a InputFile 1`] = `
<div>
  <div
    class="dropZoneBase dropZone"
    tabindex="0"
  >
    <input
      autocomplete="off"
      style="display: none;"
      tabindex="-1"
      type="file"
    />
    <div
      class="padded small"
    >
      <button
        class="button small work secondary"
        type="button"
      >
        <span
          class="base extraBold smaller uppercase green"
        >
          Upload File
        </span>
      </button>
      <p
        class="base regular small greyBlue"
      >
        or drag a file here to upload
      </p>
    </div>
  </div>
</div>
`;

exports[`renders an InputFile with multiple images allowed 1`] = `
<div>
  <div
    class="dropZoneBase dropZone"
    tabindex="0"
  >
    <input
      accept="image/*"
      autocomplete="off"
      multiple=""
      style="display: none;"
      tabindex="-1"
      type="file"
    />
    <div
      class="padded small"
    >
      <button
        class="button small work secondary"
        type="button"
      >
        <span
          class="base extraBold smaller uppercase green"
        >
          Upload Images
        </span>
      </button>
      <p
        class="base regular small greyBlue"
      >
        or drag images here to upload
      </p>
    </div>
  </div>
</div>
`;

exports[`renders an InputFile with multiple uploads 1`] = `
<div>
  <div
    class="dropZoneBase dropZone"
    tabindex="0"
  >
    <input
      autocomplete="off"
      multiple=""
      style="display: none;"
      tabindex="-1"
      type="file"
    />
    <div
      class="padded small"
    >
      <button
        class="button small work secondary"
        type="button"
      >
        <span
          class="base extraBold smaller uppercase green"
        >
          Upload Files
        </span>
      </button>
      <p
        class="base regular small greyBlue"
      >
        or drag files here to upload
      </p>
    </div>
  </div>
</div>
`;

exports[`renders an InputFile with only images allowed 1`] = `
<div>
  <div
    class="dropZoneBase dropZone"
    tabindex="0"
  >
    <input
      accept="image/*"
      autocomplete="off"
      style="display: none;"
      tabindex="-1"
      type="file"
    />
    <div
      class="padded small"
    >
      <button
        class="button small work secondary"
        type="button"
      >
        <span
          class="base extraBold smaller uppercase green"
        >
          Upload Image
        </span>
      </button>
      <p
        class="base regular small greyBlue"
      >
        or drag an image here to upload
      </p>
    </div>
  </div>
</div>
`;

exports[`renders an InputFile with proper variations 1`] = `
<div>
  <div
    class="dropZoneBase dropZone"
    tabindex="0"
  >
    <input
      autocomplete="off"
      style="display: none;"
      tabindex="-1"
      type="file"
    />
    <div
      class="padded small"
    >
      <button
        class="button small work secondary"
        type="button"
      >
        <span
          class="base extraBold smaller uppercase green"
        >
          Upload File
        </span>
      </button>
      <p
        class="base regular small greyBlue"
      >
        or drag a file here to upload
      </p>
    </div>
  </div>
  <div
    class="dropZoneBase dropZone"
    tabindex="0"
  >
    <input
      autocomplete="off"
      style="display: none;"
      tabindex="-1"
      type="file"
    />
    <div
      class="padded small"
    >
      <button
        class="button small work secondary"
        type="button"
      >
        <span
          class="base extraBold smaller uppercase green"
        >
          Upload File
        </span>
      </button>
    </div>
  </div>
  <div
    class="dropZoneBase"
    tabindex="-1"
  >
    <input
      autocomplete="off"
      style="display: none;"
      tabindex="-1"
      type="file"
    />
    <button
      class="button base work secondary fullWidth"
      type="button"
    >
      <span
        class="base extraBold small uppercase green"
      >
        Upload File
      </span>
    </button>
  </div>
  <div
    class="dropZoneBase"
    tabindex="-1"
  >
    <input
      autocomplete="off"
      style="display: none;"
      tabindex="-1"
      type="file"
    />
    <button
      class="button small work secondary fullWidth"
      type="button"
    >
      <span
        class="base extraBold smaller uppercase green"
      >
        Upload File
      </span>
    </button>
  </div>
</div>
`;
